### PR TITLE
Low-hanging Ember optimizations

### PR DESCRIPTION
**Pull request by armanbilge**
Just some micro-optimizations I noticed while working on https://github.com/http4s/http4s/pull/6752.

### CODE REVIEW

**Comment by armanbilge** [comment]
Actually the optimization I really wanted to make is to this:
https://github.com/http4s/http4s/blob/04b2bbbd1787ac7084f589240ed8a062f38ed584/ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala#L78

I was hoping we could go directly from `StringBuilder` to `Array[Byte]` without making an intermediate `String`. One way to do this is to use `CharBuffer.wrap(sb)` with the ISO_8859_1 `CharsetEncoder`. But there is an encoding fast-path for `String` that is not available for a generic `CharBuffer` so I'm skeptical if this would improve performance.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala [hunk 1]
```diff
@@ -47,31 +47,24 @@ private[ember] object Encoder {
...
-      resp.headers
-        .get[`Content-Length`]
-        .foreach { _ =>
-          appliedContentLength = true
-        }
-
       // Apply each header followed by a CRLF
       resp.headers.foreach { h =>
         if (h.isNameValid) {
+          appliedContentLength |= h.name == `Content-Length`.name
```
Comment (ChristopherDavenport) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:53]: I do not recognize this operator, is this optimized over an if with assignment. If so great, but specialized mutability stuff I don't really know.

Comment (armanbilge) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:53]: Good point. Maybe I should write it like this.
```scala
appliedContentLength = appliedContentLength || h.name == `Content-Length`.name
```

File: ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala [hunk 2]
```diff
@@ -139,21 +139,23 @@ private[ember] object ChunkedEncoding {
         .map { case (headerP, rest) => Trailers(headerP.headers, rest) }
     }
 
-  private val lastChunk: Chunk[Byte] =
-    Chunk.byteVector((ByteVector('0') ++ crlf ++ crlf).compact)
+  private[this] val lastChunk = {
```
Comment (diesalbla) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala:142]:
```suggestion
  private[this] val lastChunk: Stream[Nothing, Byte] = {
```

**Review by diesalbla — APPROVED** [review]

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala [hunk 3]
```diff
@@ -121,33 +115,25 @@ private[ember] object Encoder {
...
-        if (
-          !appliedContentLength && req.body == EmptyBody && !NoPayloadMethods.contains(req.method)
-        ) {
+        def isEmptyBody = req.body eq EmptyBody
+        def isNoPayloadMethod = NoPayloadMethods.contains(req.method)
+        if (!appliedContentLength && isEmptyBody && !isNoPayloadMethod) {
           stringBuilder.append(zeroContentLengthRaw).append(CRLF)
           chunked = false
-        } else if (!chunked && !appliedContentLength && !NoPayloadMethods.contains(req.method)) {
+        } else if (!chunked && !appliedContentLength && !isNoPayloadMethod) {
```
Comment (armanbilge) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:134]: @danicheg wdyt? I studied the bytecode and looks like these are emitted as private static methods, which I think it's safe to assume that the JIT will inline.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala [hunk 4]
```diff
@@ -121,33 +115,25 @@ private[ember] object Encoder {
...
-            ()
           }
         }
 
-        if (
-          !appliedContentLength && req.body == EmptyBody && !NoPayloadMethods.contains(req.method)
-        ) {
+        def isEmptyBody = req.body eq EmptyBody
+        def isNoPayloadMethod = NoPayloadMethods.contains(req.method)
+        if (!appliedContentLength && isEmptyBody && !isNoPayloadMethod) {
```
Comment (danicheg) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:131]: Continuing bike-shedding, speaking in terms of saving a few ns, shouldn't we place `!isNoPayloadMethod` ahead of `isEmptyBody`? I mean, `!isNoPayloadMethod` might be `false` a bit frequently then `isEmptyBody`.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala (hunk 3 above)
Comment (danicheg) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:134]: Can confirm – readability is better now :)

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala (hunk 4 above)
Comment (armanbilge) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:131]: That might be, but I think reference equality `eq` is faster than set membership `.contains`.

**Comment by armanbilge** [comment]
Apologies, since this PR is sitting I started committing additional optimizations :) each commit should be individual reviewable.

**Review by valencik — APPROVED** [review]
I think this looks good. Lots of `ByteVector.view` to avoid copying arrays and some nice `unchunks`.
It does make me think we could benefit from some more ember benchmarks. I'm going to look into how hard those might be to add.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala [hunk 5]
```diff
@@ -139,21 +139,23 @@ private[ember] object ChunkedEncoding {
         .map { case (headerP, rest) => Trailers(headerP.headers, rest) }
     }
 
-  private val lastChunk: Chunk[Byte] =
-    Chunk.byteVector((ByteVector('0') ++ crlf ++ crlf).compact)
+  private[this] val lastChunk: Stream[fs2.Pure, Byte] = {
+    val bytes = new Array[Byte](5)
+    bytes(0) = '0'.toByte
+    `\r\n\r\n`.copyToArray(bytes, 1)
+    Stream.chunk(Chunk.array(bytes))
```
[the remark this answers is not in the document]
Comment (armanbilge) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala]: Thanks yeah, that's a lot nicer 😂

**Comment by rossabaker** [comment]
Anything holding this up from merge?

**Comment by armanbilge** [comment]
Oh, when Andrew first reviewed they weren't a maintainer so I couldn't merge. Now I can!

**Pull request merged**